### PR TITLE
Collection of rebranch changes required for the compiler to build

### PR DESCRIPTION
**Pull request by bnbarham**
This set of changes enables the compiler to build, but currently asserts when building the stdlib while creating intrinsics and also when running LLVM's verifier pass.

Verifier failure:
```
Function debug format should match parent module
ptr @"$s6Darwin7FLT_MAXSfvgTm"
0
; ModuleID = '.../stdlib/swift-macosx-arm64/stdlib/public/ClangOverlays/IOS/arm64e/_Builtin_float.o'
1
```
ie. the module has `IsNewDbgInfoFormat` set but the function does not.

Intrinsic assertion:
```
Assertion failed: (Index < Length && "Invalid index!"), function operator[], file ArrayRef.h, line 257.
...
3.	While evaluating request IRGenRequest(IR Generation for module Synchronization)
4.	While emitting IR SIL function "@$s15Synchronization19AtomicLazyReferenceVfD".
 for 'deinit' (at .../swift/stdlib/public/Synchronization/AtomicLazyReference.swift:31:3)
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
...
5  libsystem_c.dylib        0x0000000183dc9a40 abort + 180
6  libsystem_c.dylib        0x0000000183dc8d30 err + 0
7  swift-frontend           0x00000001074b942c DecodeFixedType(llvm::ArrayRef<llvm::Intrinsic::IITDescriptor>&, llvm::ArrayRef<llvm::Type*>, llvm::LLVMContext&) (.cold.4) + 0
8  swift-frontend           0x0000000105ee1784 DecodeFixedType(llvm::ArrayRef<llvm::Intrinsic::IITDescriptor>&, llvm::ArrayRef<llvm::Type*>, llvm::LLVMContext&) + 744
9  swift-frontend           0x0000000105ee1340 llvm::Intrinsic::getType(llvm::LLVMContext&, unsigned int, llvm::ArrayRef<llvm::Type*>) + 116
10 swift-frontend           0x0000000105ee33c8 llvm::Intrinsic::getDeclaration(llvm::Module*, unsigned int, llvm::ArrayRef<llvm::Type*>) + 48
11 swift-frontend           0x000000010088b470 swift::irgen::IRBuilder::CreateIntrinsicCall(unsigned int, llvm::ArrayRef<llvm::Value*>, llvm::Twine const&) + 84
...
```
Possibly related to the above?

### CODE REVIEW

File: include/swift/SIL/SILInstruction.h [hunk 1]
```diff
@@ -94,6 +94,7 @@ template <> struct compute_node_options<::swift::SILInstruction> {
 
     static const bool enable_sentinel_tracking = false;
     static const bool is_sentinel_tracking_explicit = false;
+    static const bool has_iterator_bits = false;
```
Comment (bnbarham) [inline include/swift/SIL/SILInstruction.h:97]: @felipepiovezan I assume we probably want to have an iterator bit here?

File: lib/IRGen/IRGenDebugInfo.cpp [hunk 2]
```diff
@@ -3350,21 +3350,21 @@ struct DbgIntrinsicEmitter {
...
-                            llvm::Instruction *InsertBefore) {
+  llvm::DbgInstPtr insert(llvm::Value *Addr, llvm::DILocalVariable *VarInfo,
+                          llvm::DIExpression *Expr,
+                          const llvm::DILocation *DL,
+                          llvm::Instruction *InsertBefore) {
     if (ForceDbgDeclare == AddrDbgInstrKind::DbgDeclare)
       return DIBuilder.insertDeclare(Addr, VarInfo, Expr, DL, InsertBefore);
     Expr = llvm::DIExpression::append(Expr, llvm::dwarf::DW_OP_deref);
     return DIBuilder.insertDbgValueIntrinsic(Addr, VarInfo, Expr, DL,
                                             InsertBefore);
```
Comment (bnbarham) [inline lib/IRGen/IRGenDebugInfo.cpp:3361]: @felipepiovezan  I assume this is the underlying cause of the crashes and that we should instead use the new (old) intrinsic functions (https://github.com/llvm/llvm-project/blob/main/llvm/docs/RemoveDIsDebugInfo.md):
```
LLVMDIBuilderInsertDeclareIntrinsicBefore   # Insert a debug intrinsic (old debug info format).
LLVMDIBuilderInsertDeclareIntrinsicAtEnd    # Same as above.
LLVMDIBuilderInsertDbgValueIntrinsicBefore  # Same as above.
LLVMDIBuilderInsertDbgValueIntrinsicAtEnd   # Same as above.
```

**Comment by bnbarham** [comment]
@beccadax / @egorzhdan I'd appreciate a review on the ClangImporter changes. @aschwaighofer could you take a look over the IRGen ones? And @eeckstein for the SIL/Opt changes. They're all generally pretty small.

@jansvoboda11 I don't think there was anything particularly crazy with the `FileEntryRef` changes, but a second pair of eyes would be nice. Also the OptTable macros, but again, that seemed pretty straight forward (also, thank you, that is *so* much better now).

**Review by egorzhdan — APPROVED** [review]
ClangImporter changes LGTM, thanks!

**Review by jansvoboda11 — APPROVED** [review]
The `FileEntry` to `FileEntryRef` changes LGTM. So does usage of the new `LLVM_*_OPT_*` macros, but note that you might need to ensure you use the `/Zc:preprocessor` flag with MSVC (and maybe bump the minimal supported version): https://reviews.llvm.org/D135128

**Comment by bnbarham** [comment]
>  but note that you might need to ensure you use the `/Zc:preprocessor` flag with MSVC

Ooo thanks, good to know!

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
Going to go ahead and merge to continue along. Thanks for reviewing @jansvoboda11 and @egorzhdan!

@aschwaighofer / @eeckstein when you have the time I'd really appreciate a quick look over 🙏

**Pull request merged**